bpf: snat: DSR-eligible traffic can skip check for Nodeport NAT conflict

julianwiedmann opened the pull request:
Right now the SNAT logic in `to-overlay` / `to-netdev` tracks all (host-originating) connections that potentially conflict with NAT connections by the N/S LB to remote backends (ie. `tail_nodeport_nat_egress_ipv4()`).

If the LB uses DSR for the packet's L4 protocol type, there's no risk of conflicting with a NATed LB connection. So we don't need to track such a connection.

julianwiedmann committed:
bpf: nat: split up Nodeport NAT and Masquerading checks

Right now we're mixing all "is SNAT needed?" checks into one big function.
Split them up into two parts, so it's easier to understand why they're
needed:
1) conflicts vs Nodeport NAT: when host-originating connections use the
   same source IP as NATed Nodeport connections, we need to prevent
   conflicts with the Nodeport NAT port range.
2) Masquerading: besides the obvious reason (pod traffic leaving the
   cluster, potentially targeted by an EgressGW policy), we also need
   to prevent conflicts with host-originating connections that use the
   selected Masquerading IP(s).

We'll limit the checks for Nodeport NAT conflicts in a subsequent patch.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: nat: extract CT tuple earlier in SNAT path

Extract the "easy" parts of the CT tuple (everything but the L4 ports)
earlier in the SNAT flow. This makes for easier error handling, and
provides access to the L4 protocol during the Nodeport NAT-conflict check
(where we will use it in a subsequent patch).

In the future we can also pipe this into the snat_v*_needs_masquerade()
check, where a CT tuple is required as well.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: nodeport: only check SVC protocols for NAT conflicts

When a host-originating connection uses a L4 protocol that is not a valid
SVC protocol (ie. it's not TCP / UDP / SCTP), we know that it won't
conflict with a NATed Nodeport connection to the same destination. Thus we
can skip the corresponding check in the SNAT path.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

gentoo-root commented on bpf/lib/nodeport.h:
```diff
@@ -1504,13 +1516,21 @@ static __always_inline int nodeport_snat_fwd_ipv4(struct __ctx_buff *ctx,
 	if (!revalidate_data(ctx, &data, &data_end, &ip4))
 		return DROP_INVALID;
 
+	tuple.flags = TUPLE_F_OUT;
```
Nit: why not use snat_v4_init_tuple?

julianwiedmann replied:
Hmm can't think of a good reason. Thanks, that also reduces the churn in the tests quite a bit.

gentoo-root commented on bpf/lib/nodeport.h:
```diff
@@ -1424,6 +1465,32 @@ static __always_inline bool nodeport_uses_dsr4(const struct ipv4_ct_tuple *tuple
+#endif /* TUNNEL_MODE && IS_BPF_OVERLAY */
+
+#if defined(IS_BPF_HOST)
+    /* NATIVE_DEV_IFINDEX == DIRECT_ROUTING_DEV_IFINDEX cannot be moved into
```
Nit: tab for indent.

julianwiedmann replied:
Ack. While at it I also added a temporary variable that holds `DIRECT_ROUTING_DEV_IFINDEX` (to appease checkpatch).

gentoo-root commented on bpf/lib/nodeport.h:
```diff
@@ -1531,7 +1532,8 @@ static __always_inline int nodeport_snat_fwd_ipv4(struct __ctx_buff *ctx,
 	l4_off = ETH_HLEN + ipv4_hdrlen(ip4);
 
-	if (nodeport_has_nat_conflict_ipv4(ip4, &target))
+	if (lb_is_svc_proto(tuple.nexthdr) && !nodeport_uses_dsr4(&tuple) &&
```
Could you please elaborate why we skip apply_snat for ICMP?

julianwiedmann replied:
Thanks for pointing that out :). I should do even better, and split that change off into a separate patch. Done so now.

We're not forwarding ICMP packets from the LB to a backend, so there's no risk of conflicting with such traffic.

julianwiedmann committed:
bpf: nodeport: only check for NAT conflict when not using DSR

When a host-originating connection uses a L4 protocol that is DSR-eligible,
we know that it won't conflict with a NATed Nodeport connection to the
same destination (as that connection would use DSR, not NAT). Thus we
can skip the corresponding check in the SNAT path.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>
